### PR TITLE
Propagate container across object cast

### PR DESCRIPTION
**Pull request by pnkfelix**
Given `<expr> as Box<Trait>`, infer that `Box<_>` is expected type for `<expr>`.

This is useful for addressing fallout from newly proposed box protocol; see #22006 for examples of such fallout, much of which will be unnecessary with this fix.

### CODE REVIEW

**Comment by pnkfelix** [comment]
cc @nikomatsakis

File: src/test/run-pass/infer-container-across-object-cast.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// except according to those terms.
+
+// Given `<expr> as Box<Trait>`, we should be able to infer that a
+// `Box<_>` is the expected type.
+
+trait Foo { fn foo(&self) -> u32; }
+impl Foo for u32 { fn foo(&self) -> u32 { *self } }
+
+trait Boxed { fn make() -> Self; }
+impl Boxed for Box<u32> { fn make() -> Self { Box::new(7) } }
```
Comment (eddyb) [inline src/test/run-pass/infer-container-across-object-cast.rs]: Could you add a second `impl` that returns something which is not `Box<_>`? (same for `Refed` below)
I recall being bit before by having just one `impl` and getting it selected even without enough type hints.

Comment (pnkfelix) [inline src/test/run-pass/infer-container-across-object-cast.rs]: Will do.  (Its interesting, @alexcrichton brought up the same issue on the RFC itself.  I wonder if this is an artifact of the old trait matching code, or the new, or both...)

**Comment by eddyb** [comment]
r=me with the test additions.

**Comment by nrc** [comment]
Are we going to miss this? I have a branch which removes this facility. The root problem is that we do object conversions differently depending on if we are using coercion or casting. My branch removes the cast code path completely so `x as Box<Foo>` is essentially compiled as `(x: Box<Foo>) as Box<Foo>`. The solution is to re-implement this patch for coercions, but I'm not sure if that is possible because coercions are a little less 'directed' by the types involved.

**Comment by nrc** [comment]
ping @pnkfelix   about that last comment

**Comment by pnkfelix** [comment]
@nrc when you say "so x as Box<Foo> is essentially compiled as `(x: Box<Foo>) as Box<Foo>`", you mean that is what you do on your branch?

If so, then I think that should be fine; the important thing was getting the type-hint that `x` has some `Box<_>` type, at least for my purposes.

If you look at [appendix b](https://github.com/pnkfelix/rfcs/blob/fsk-box-and-in-for-stdlib-rfc/text/0000-box-and-in-for-stdlib.md#appendix-b-examples-of-interaction-between-desugaring-type-inference-and-coercion) from the most recent box/placement RFC, it shows the particular place where I needed this PR, namely the bit that says:

```rust
#[cfg(coerce_works3)] // (This one assumes PR 22012 has landed)
pub fn coerce<'a, F>(f: F) -> BoxFn<'a> where F: Fn(), F: 'a { box_!( f ) as BoxFn }
```

so you should be able to take that code, put it into a file, and try your branch on it, to make sure you don't regress the behavior here.  (I guess I should have maybe put in some tests to ensure this did not get regressed.)

In other words, make sure that this [playpen](http://is.gd/cuJPL7) does not stop working, and I'll be happy.

**Comment by pnkfelix** [comment]
@nrc and I discussed further; I was over-optimistic when I said:
> If so, then I think that should be fine
because what I would have really liked would be to compile `x as Box<Foo>` into `(x: Box<_>) as Box<Foo>` (and *not* `(x: Box<Foo> as Box<Foo>)`)

but in any case, after further discussion I decided that its not the end of the world if @nrc makes the change he desires here.
